crates/astria-sequencer-validation/src/proof.rs

```diff
@@ -0,0 +1,214 @@
+use ct_merkle::{
+    inclusion::InclusionProof as CtInclusionProof,
+    CtMerkleTree,
+    RootHash,
+};
+use serde::{
+    ser::SerializeStruct,
+    Deserialize,
+    Serialize,
+};
+use sha2::Sha256;
+use tendermint::Hash;
+
+/// A wrapper around [`ct_merkle::CtMerkleTree`], which uses sha256 as the hashing algorithm
+/// and Vec<u8> as the leaf type.
+#[derive(Debug, Default)]
+pub struct MerkleTree(CtMerkleTree<Sha256, Vec<u8>>);
+
+impl MerkleTree {
+    #[must_use]
+    pub fn from_leaves(leaves: Vec<Vec<u8>>) -> Self {
+        let tree = leaves
+            .into_iter()
+            .fold(CtMerkleTree::new(), |mut tree, leaf| {
+                tree.push(leaf);
+                tree
+            });
+        MerkleTree(tree)
+    }
+
+    #[must_use]
+    pub fn root(&self) -> Hash {
+        let digest = sha2::digest::Output::<Sha256>::from_slice(self.0.root().as_bytes())
+            .as_slice()
+            .to_vec();
+        Hash::from_bytes(tendermint::hash::Algorithm::Sha256, &digest)
+            .expect("cannot fail since both hashes are 32 bytes")
+    }
+
+    /// Returns the inclusion proof for the leaf at the given index.
+    ///
+    /// # Errors
+    ///
+    /// - if the index is out of bounds
+    pub fn prove_inclusion(&self, idx: usize) -> eyre::Result<InclusionProof> {
+        Ok(InclusionProof {
+            value: self
+                .0
+                .get(idx)
+                .ok_or(eyre::eyre!("index out of bounds of merkle tree"))?
+                .clone(),
+            idx,
+            num_leaves: self.0.len(),
+            inclusion_proof: self.0.prove_inclusion(idx),
+        })
+    }
+}
+
+/// A merkle proof of inclusion.
+#[derive(Debug)]
+#[allow(clippy::module_name_repetitions)]
+pub struct InclusionProof {
```

Comment:
Am I right that the main purpose of this type is to allow sending proofs over the wire? Should the value-to-be-proven indeed be serialized along with the proof?

Reply:
I actually removed the value from here in the follow up, will update here also. yes the type basically only has Serialize, Deserialize, and verify functions